src/issuance_auth.rs

```diff
@@ -0,0 +1,443 @@
+    ///
+    /// [ZIP 32]: https://zips.z.cash/zip-0032
+    pub fn random(rng: &mut impl CryptoRngCore) -> Self {
+        Self(NonZeroScalar::random(rng))
+    }
+
+    fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes().to_vec()
+    }
+
+    /// Deserialize the issuance authorization signature from its canonical byte representation.
+    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
+        NonZeroScalar::try_from(bytes).ok().map(Self)
+    }
+
+    /// Derives the Orchard-ZSA issuance key for the given seed, coin type, and account.
+    pub fn from_zip32_seed(
+        seed: &[u8],
+        coin_type: u32,
+        account: u32,
+    ) -> Result<Self, zip32::Error> {
+        if account != 0 {
+            return Err(zip32::Error::NonZeroAccount);
+        }
+
+        // Call zip32 logic
+        let path = &[
+            ChildIndex::hardened(ZIP32_PURPOSE_FOR_ISSUANCE),
+            ChildIndex::hardened(coin_type),
+            ChildIndex::hardened(account),
+        ];
+
+        // we are reusing zip32 logic for deriving the key, zip32 should be updated as discussed
+        let &isk_bytes = ExtendedSpendingKey::<zip32::Issuance>::from_path(seed, path)?
+            .sk()
+            .to_bytes();
+
+        Self::from_bytes(&isk_bytes).ok_or(zip32::Error::InvalidSpendingKey)
+    }
+
+    /// Signs a 32-byte message using the issuance authorizing key.
+    ///
+    /// Only supports signing of messages of length 32 bytes, since we will only be using it
+    /// to sign 32 byte SIGHASH values.
+    pub fn try_sign(&self, msg: &[u8; 32]) -> Result<IssueAuthSig<ZSASchnorr>, Error> {
+        ZSASchnorr::try_sign_with_aux(&self.0, msg, &[0u8; 32]).map(IssueAuthSig::<ZSASchnorr>)
+    }
+}
+
+impl Debug for IssueAuthKey<ZSASchnorr> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        // Do not print bytes from the issuance authorizing key.
+        let ik = IssueValidatingKey::from(self);
+        write!(f, "IssueAuthKey({:?})", ik)
+    }
+}
+
+impl From<&IssueAuthKey<ZSASchnorr>> for IssueValidatingKey<ZSASchnorr> {
+    fn from(isk: &IssueAuthKey<ZSASchnorr>) -> Self {
+        Self(*schnorr::SigningKey::from(isk.0).verifying_key())
+    }
+}
+
+impl IssueValidatingKey<ZSASchnorr> {
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes().to_vec()
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Option<Self> {
+        VerifyingKey::from_bytes(bytes).ok().map(Self)
+    }
+
+    /// Encodes the issuance validating key into a byte vector, in the manner defined in [ZIP 227][issuancekeycomponents].
+    ///
+    /// [issuancekeycomponents]: https://zips.z.cash/zip-0227#derivation-of-issuance-validating-key
+    pub fn encode(&self) -> Vec<u8> {
+        let ik_bytes = self.to_bytes();
+        let mut encoded = Vec::with_capacity(1 + ik_bytes.len());
+        encoded.push(ZSASchnorr::ALGORITHM_BYTE);
+        encoded.extend(ik_bytes);
+        encoded
```

Comment:
same:
```rust
    let mut encoded = vec![ZSASchnorr::ALGORITHM_BYTE];
    encoded.extend(self.to_bytes());
    encoded
```
(all)

Reply:
I prefer to use with_capacity because it preallocates the exact size avoiding reallocations.
In addition extend_from_slice copies bytes in one shot which is faster than extend.

Reply:
If so, use size of `ZSASchnorr::ALGORITHM_BYTE` + `ik_bytes.len()`, not `+1`